### ISSUE TITLE
Support for `restrict`

### ISSUE THREAD

**Issue by ch3root** [TrustInSoft/tis-interpreter#69]
Source code:
```c
#include <stdio.h>

int f(int *restrict p, int *q)
{
*p = 1;
*q = 2;
  return *p;
}

int main()
{
  int x;

  printf("%d\n", f(&x, &x));
}

```

tis-interpreter (274be6de79973f11ab62a3c77d0981882734cb2d) output:
```
[value] Analyzing a complete application starting at main
[value] Computing initial state
[value] Initial state computed

2

[value] done for function main
```

gcc 7 w/ubsan output:
```
1
```

clang 3.9 w/ubsan output:
```
restrict.c:3:5: warning: no previous prototype for function 'f' [-Wmissing-prototypes]
int f(int *restrict p, int *q)
    ^
1 warning generated.
2
```

I don't think silently ignoring `restrict` is a good idea. Please note that the gcc output differs.

**Comment by pascal-cuoq** [comment]
`restrict` is [not](http://blog.frama-c.com/index.php?post/2012/08/02/restrict-not-modular) my favorite C99 addition.

Concretely, a `restrict` analysis cannot warn for `f(&x, &x)`: nothing illegal has happened in the program so far. It cannot warn for `*p = 1;`: the function has not accessed `x` through another access path yet. It is only when it encounters `*q = 2;` that it must remember that `p` is a `restrict` pointer and that the location being accessed was also accessed through `p`.

To whomever is reading this: if you are looking for a project or internship, this analysis is up for grabs. Send me an e-mail.

**Comment by ch3root** [comment]
I disagree with the reasoning in your blogpost but I understand that this is not an easy feature to implement. My main point is not that `restrict` is not fully analyzed. It's that there is no indication that `restrict` is not analyzed.

Right now tis-interpreter gives results different from gcc without any warnings (and no gcc bugs involved). "[useful theorem](https://github.com/TrustInSoft/tis-interpreter/issues/43#issuecomment-214915655)" and such...

Given that it's not clear if `restrict` will be properly supported I see two solutions:
- add a warning on the first occurred `restrict` about unsoundness of further analysis;
- remove support for `restrict` completely and recommend a user to just `#define restrict` if needed.
